include/jemalloc/internal/tsd_types.h

```diff
@@ -291,7 +291,8 @@ a_name##tsd_cleanup_wrapper(void) {					\
 	if (wrapper == NULL) {						\
 		return false;						\
 	}								\
-	if (a_cleanup != malloc_tsd_no_cleanup &&			\
+	if (/* a_cleanup != malloc_tsd_no_cleanup && */			\
+	    /* Verify compiler (mingw32+gcc) bug.*/			\
```

Comment:
Regarding workarounds to avoid the compiler issue:

1) I tried replacing malloc_tsd_no_cleanup with NULL, i.e. changing "a_cleanup != malloc_tsd_no_cleanup" to "a_cleanup != NULL", and that worked fine, except that a bunch of warnings about the comparison will always be as the address will never be NULL. (I suspect this was part of the reason to have malloc_tsd_no_cleanup?)
2) keep malloc_tsd_no_cleanup (change it to an empty function instead of assert not_reached), but remove the check "a_cleanup != malloc_tsd_no_cleanup". Theoretically compiler should still be able to optimize this away when malloc_tsd_no_cleanup is actually used. 

Thoughts @davidtgoldblatt?